java-checks/src/main/java/org/sonar/java/checks/security/UserEnumerationCheck.java

```diff
@@ -0,0 +1,123 @@
+  private static final String USER_DETAILS_SERVICE = "org.springframework.security.core.userdetails.UserDetailsService";
+  private static final String USERNAME_NOT_FOUND_EXCEPTION = "org.springframework.security.core.userdetails.UsernameNotFoundException";
+  private static final String HIDE_USER_NOT_FOUND_EXCEPTIONS = "setHideUserNotFoundExceptions";
+  private static final String LOAD_USER_BY_USERNAME = "loadUserByUsername";
+  private static final String BOOLEAN = "boolean";
+  private static final String STRING = "java.lang.String";
+  private static final String THROWABLE = "java.lang.Throwable";
+
+  private final MethodMatchers setHideUserMatcher = MethodMatchers.create()
+    .ofSubTypes(ABSTRACT_USER_DETAILS_AUTHENTICATION_PROVIDER)
+    .names(HIDE_USER_NOT_FOUND_EXCEPTIONS)
+    .addParametersMatcher(BOOLEAN)
+    .build();
+
+  private final MethodMatchers loadUserMatcher = MethodMatchers.create()
+    .ofSubTypes(USER_DETAILS_SERVICE)
+    .names(LOAD_USER_BY_USERNAME)
+    .addParametersMatcher(STRING)
+    .build();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Arrays.asList(Tree.Kind.METHOD_INVOCATION, Tree.Kind.THROW_STATEMENT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree.is(Tree.Kind.THROW_STATEMENT)) {
+      checkThrowUsernameNotFoundException(((ThrowStatementTree) tree));
+      return;
+    }
+
+    MethodInvocationTree methodInvocationTree = (MethodInvocationTree) tree;
+    Arguments arguments = methodInvocationTree.arguments();
+    if (arguments.isEmpty()){
+      return;
+    }
+    ExpressionTree expression = arguments.get(0);
+
+    checkHiddenUserNotFoundException(methodInvocationTree, expression);
+
+    checkLoadUserArgUsedInExceptions(methodInvocationTree, expression);
+  }
+
+  private void checkLoadUserArgUsedInExceptions(MethodInvocationTree methodInvocationTree, ExpressionTree expression) {
+    if (loadUserMatcher.matches(methodInvocationTree) && expression.is(Tree.Kind.IDENTIFIER)) {
+      IdentifierTree identifierTree = (IdentifierTree) expression;
+      identifierTree.symbol().usages()
+        .stream().filter(UserEnumerationCheck::checkParentIsThrowable)
+        .forEach(value -> reportIssue(value, MESSAGE));
+    }
+  }
+
+  private void checkHiddenUserNotFoundException(MethodInvocationTree methodInvocationTree, ExpressionTree expression) {
+    if (setHideUserMatcher.matches(methodInvocationTree) && isFalseLiteral(expression)) {
+      reportIssue(methodInvocationTree, MESSAGE);
+    }
+  }
+
+  private void checkThrowUsernameNotFoundException(ThrowStatementTree tree) {
+    if (tree.expression().symbolType().is(USERNAME_NOT_FOUND_EXCEPTION)) {
+      reportIssue(tree, MESSAGE);
+    }
+  }
+
+  private static boolean isFalseLiteral(ExpressionTree expression) {
+    return expression.is(Tree.Kind.BOOLEAN_LITERAL) && !Boolean.parseBoolean(((LiteralTree) expression).value());
```

Comment:
What about using `LiteralUtils.isFalse(tree)` here?
I also feel that we might be missing cases of constants, such as:
```
daoauth.setHideUserNotFoundExceptions(MY_CONSTANT);
```
From an expressionTree, you can probably call `asConstant(Boolean.class)`, and it might work.